feat: Organizational Chart

ruchamahabal opened the pull request:
## Organizational Charts

An organizational chart shows the organizational structure by depicting connections between different employees with their designation, image, and name. In large organizations where it's difficult to remember names, people can interact with the org chart and know the hierarchy through faces and other info. The **Edit** button navigates to the employee master.

## Desktop View

### Interactions:

https://user-images.githubusercontent.com/24353136/124869278-975b2a00-dfde-11eb-8c91-d067e6ce94b9.mp4

### Expand All and Export

1. Option to expand all the nodes. Interaction is disabled until you collapse all.
2. Export to PNG

https://user-images.githubusercontent.com/24353136/126934843-9d1b100d-054c-4a3e-8175-752d832be0cd.mp4

## Mobile View

![image](https://user-images.githubusercontent.com/24353136/126277752-91f84a56-7485-4dfb-a295-447380c5eaea.png)

![Untitled 2](https://user-images.githubusercontent.com/24353136/123870322-cbd04580-d94f-11eb-9e01-8f9015871d6a.png)

After clicking on sibling group:

![node-group](https://user-images.githubusercontent.com/24353136/123870674-4dc06e80-d950-11eb-8dc0-e9ad23bbee14.png)

Documentation PR: https://github.com/frappe/erpnext_documentation/pull/364

ToDo

- [x] Optimise rendering
- [x] Handle multiple root nodes / orphan nodes
- [x] Expand all nodes
- [x] Export Chart
- [x] Tests
- [x] Documentation

ruchamahabal committed:
feat: Organizational Chart

ruchamahabal committed:
feat: org chart mobile interactions

ruchamahabal committed:
feat(mobile): sibling node group expansion and rendering

ruchamahabal committed:
fix: expanded node group interactions and visibility

ruchamahabal committed:
feat: connectors for mobile node cards

ruchamahabal committed:
fix: don't refresh connections for same node

- remove all connectors while expanding a group node

ruchamahabal committed:
chore: create separate files for Desktop and Mobile view and bundle assets

ruchamahabal committed:
refactor: add options to chart

- method to return the node data

- wrapper for showing the hierarchy

ruchamahabal committed:
feat: setup node edit action

ruchamahabal committed:
fix: revert changes in employee descendants query

ruchamahabal committed:
refactor: use arcs instead of bezier curves for cleaner connectors

ruchamahabal committed:
feat: add arc to connectors in mobile view

ruchamahabal committed:
fix: edit node button overflowing

ruchamahabal committed:
fix: sider

olamide2 commented:
This is Great

Will we be able to change relationship by using  drag and drop from the chart ?

ruchamahabal commented:
> This is Great
> 
> Will we be able to change relationship by using drag and drop from the chart ?

Thanks. No this is just for visualising the hierarchy based on the "reports to" field and not building the employee tree.

ruchamahabal committed:
fix: removing orphaned connectors

ruchamahabal committed:
fix: unnecessary variables

ruchamahabal committed:
feat: handle multiple root / orphan nodes

ruchamahabal committed:
perf: Optimise Rendering

- optimise get_children function

- use promises instead of callbacks

- optimise selectors

- use const wherever possible

- use pure js instead of jquery for connectors for faster rendering

ruchamahabal committed:
fix: do not sort by number of connections

ruchamahabal committed:
feat: use icon for connections on mobile view

ruchamahabal committed:
fix: exclude active node while fetching sibling group

ruchamahabal committed:
fix: sibling group expansion not working for root nodes

ruchamahabal committed:
fix(mobile): collapsed nodes not expanding

ruchamahabal committed:
fix: sider

ruchamahabal committed:
Merge branch 'version-13-hotfix' of https://github.com/frappe/erpnext into org-chart

ruchamahabal committed:
test: introduce cypress tests

Co-authored-by: Ankush <ankush@iwebnotes.com>
Co-authored-by: Nabin Hait <nabinhait@gmail.com>

ruchamahabal committed:
test: UI tests for org chart desktop

ruchamahabal committed:
test: UI tests for org chart mobile

fix(mobile): detach node before emptying hierarchy

fix(mobile): sibling group not rendering for first level

ruchamahabal committed:
fix: sider

ruchamahabal committed:
ci(cypress): use env variable for key

documentation ref: https://docs.cypress.io/guides/guides/command-line\#cypress-run

ruchamahabal committed:
fix: UI tests

ruchamahabal committed:
fix(tests): clear filter before typing

ruchamahabal committed:
fix(tests): apply filters correctly

ruchamahabal committed:
fix: tests

ruchamahabal committed:
fix: tests

ruchamahabal committed:
fix: tests

ruchamahabal committed:
fix(test): increase timeout for record creation

ruchamahabal committed:
fix: sider

ruchamahabal committed:
fix: sider

ruchamahabal committed:
feat: Expand All nodes option in Desktop view

aschl commented:
Closes https://github.com/frappe/erpnext/issues/4087 and https://github.com/frappe/erpnext/issues/18819

ruchamahabal committed:
feat: add html2canvas for easily exporting html to images using canvas

ruchamahabal committed:
feat: Export chart option in desktop view

ruchamahabal committed:
fix(style): longer titles overflowing

ruchamahabal committed:
fix: remove unnecessary imports

ruchamahabal committed:
fix: test

nabinhait commented on .github/helper/install.sh:
```diff
@@ -42,5 +42,5 @@ sed -i 's/socketio:/# socketio:/g' Procfile
 
 bench get-app erpnext "${GITHUB_WORKSPACE}"
-bench start &
+bench start &> bench_run_logs.txt &
```
Should it be `bench start > bench_run_logs.txt &`

ankush replied:
I added this to log the bench console which can be shown in case of some failure. 

`&>` is for capturing both stdout and stderr.
